ring/func/sync.py

```diff
@@ -205,11 +206,82 @@ def touch_value(self, key, expire):
+        assert round(target_size) > 0, 'target_size has to be at least 1'
+        assert expire_f is None or callable(expire_f), 'expire_f has to be function or None'
+        assert isinstance(backend, type({})), 'backend has to be dict-like'
+        self._backend = backend
+        self._target_size = round(target_size)
+        self._expire_f = expire_f
+        self._mutex = threading.Lock()
+
+    def run(self):
+        class WorkThread(threading.Thread):
+            DEBUG = False
+
+            def __init__(self, outer_instance):
+                threading.Thread.__init__(self)
+                self._backend = outer_instance._backend
+                self._target_size = outer_instance._target_size
+                self._expire_f = outer_instance._expire_f
+                self._mutex = outer_instance._mutex
+
+            def strategy_with_expire(self):
+                MAX_EXPIRE_RETRY_COUNT = 4
+                now = time.time()
+                retry_count = 0
+                keys = list(self._backend.keys())
+                while (len(self._backend) > self._target_size) and (retry_count < MAX_EXPIRE_RETRY_COUNT):
+                    k = keys.pop(random.randrange(len(keys)))
+                    val = self._backend.get(k, None)
+                    if val is None:
+                        continue
+                    expire = self._expire_f(val)
+                    if expire < now:
+                        self._backend.pop(k, None)
+                        self.print_d('{} removed from strategy_with_expire => size {}'.format(k, len(self._backend)))
+                    else:
+                        retry_count += 1
+                    if len(keys) == 0:
+                        keys = list(self._backend.keys())
+
+            def strategy_with_force(self):
+                keys = list(self._backend.keys())
+                while (len(self._backend) > self._target_size) and len(keys) > 0:
+                    k = keys.pop(random.randrange(len(keys)))
+                    self._backend.pop(k, None)
+                    self.print_d('{} removed from strategy_with_force => size {}'.format(k, len(self._backend)))
+                    if len(keys) == 0:
+                        keys = list(self._backend.keys())
+
+            def run(self):
+                try:
+                    self.print_d('gc started in size:{}'.format(len(self._backend)))
+                    if self._expire_f is not None:
+                        self.strategy_with_expire()
+                    self.strategy_with_force()
+                    self.print_d('gc ended in size:{}'.format(len(self._backend)))
+                finally:
+                    self._mutex.release()
+
+
+            def print_d(self, str_):
+                if self.DEBUG:
+                    print(str_)
+
+        self._mutex.acquire()
+        if (len(self._backend) > self._target_size):
+            WorkThread(self).start()
```

Comment:
Do we take any advantage by using threads? I think gc for dict is cpu-bouded job and there is no benefits by threading. Tell me if I missed something.

Reply:
Yes, you're right. The `cache` itself is in-memory so won't be any heavy I/O issue. However, If the `dict` entries are big enough, like 1000000, `set_value`(or in some other methods) can be blocked quite an amount of time. At least we can context switch if we go multithreading. 
This is just my opinion and have no issues removing the threading.

Reply:
Let's remove it. If the size is the matter, I think there are 2 reasons to regard threading is overkill.

1. If the backend requires IO job, it makes sense. But for dict, even 1000000 iteration is a microsecond problem. In most of cases with reasonable functions, it will be less than 1000000.
2. If we correctly add locking to it, gc blocks any kind of adding and removing from the dict - which mean the caller still should wait for gc job.

Reply:
Got it!